### DISCUSSION TITLE
CulturedFact in public API !?

### DISCUSSION THREAD

**Discussion by lg2de** [xunit/xunit#2182] [Ideas]
I would like to use the CulturedFact (and related) in my own project(s).
Can we make these classes public?

Do you think it could become public part of xUnit 3?
For xUnit 2.4 a separate package could/should be created.

If you agree I could create such project with your existing code.

**Comment by bradwilson** [comment]
Yes, this is something I'll consider for v3.

**Reply by lg2de** [reply]
Do you think it will be integrated or separate project. 
Depending on the schedule of v3 I would prefer separate library, already now compatible for v2.
If you agree I could create such project/nuget package WITH OUR CODE.

**Reply by bradwilson** [reply]
We would not do an external project for v2, as we have discontinued all work on v2.

**Reply by lg2de** [reply]
V3 is out now. I could not CulturedFact in the public API.
Do you still consider it for the future?

**Reply by bradwilson** [reply]
Honestly I had forgotten about this. Yes, I can put this into the next v3 release.

**Comment by bradwilson** [comment]
Available in v3 `3.0.1-pre.15` https://xunit.net/docs/using-ci-builds

**Reply by bradwilson** [reply]
There was one change from the version we used in our tests: there are no default cultures. Our defaults (`"en-US"` and `"fr-FR"`) are not necessarily applicable to everybody else.

The attributes have been left unsealed, if you'd prefer to make versions of the attributes with your own default cultures.

**Reply by lg2de** [reply]
I've tested and it works as expected. Thanks a lot!
But...
While searching (again) for current alternatives I found a hint to one of your examples: https://github.com/xunit/samples.xunit/blob/main/v3/UseCultureExample/UseCultureAttribute.cs
(This was just after my questions yesterday and your answer.)

Looking at that and other extensions, such as https://www.nuget.org/packages/Xunit.StaFact, I'm wondering if it would be better to provide the functionality in an additional attribute, similar to what the `UseCultureAttribute` does. This would allow the combination of, for example, `WpfFact` with `UseCulture`, which I was looking for in the past.

The only thing all solutions should address is race conditions with other tests when tests are executed in parallel.

**Reply by bradwilson** [reply]
`UseCulture` can be a separate attribute because it only supports a single culture (which can be done by just modifying the behavior of the existing test case execution). `CulturedFact` supports multiple cultures, which requires being a `Fact` replacement (since it must generate multiple test cases).

If you only want a single culture (and need to use an already custom Fact attribute), then `UseCulture` would be a better choice for your usage.

**Reply by lg2de** [reply]
Yeah, that's right.
Do you plan to add UseCulture to the public API?
And what do you think about concurrent tests?

**Reply by bradwilson** [reply]
There are no plans for `UseCulture` to become part of the core. For the vast majority of users, `CulturedFact` is sufficient, and adding a variant just adds confusion about which way is "best".

There are several issues open that mention parallelism (concurrency), maybe whatever your question or request is can be addressed in one of them: https://github.com/xunit/xunit/issues?q=is%3Aissue%20state%3Aopen%20parallel
